pinot-core/src/test/java/org/apache/pinot/core/upsert/PartitionUpsertMetadataManagerTest.java

```diff
@@ -0,0 +1,199 @@
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertNull;
+import static org.testng.Assert.assertSame;
+
+
+public class PartitionUpsertMetadataManagerTest {
+  private static final String SEGMENT_PREFIX = "testSegment";
+
+  @Test
+  public void testAddSegment() {
+    PartitionUpsertMetadataManager upsertMetadataManager = new PartitionUpsertMetadataManager();
+    Map<PrimaryKey, RecordLocation> recordLocationMap = upsertMetadataManager._primaryKeyToRecordLocationMap;
+
+    // Add the first segment
+    String segment1 = SEGMENT_PREFIX + 1;
+    List<RecordInfo> recordInfoList1 = new ArrayList<>();
+    recordInfoList1.add(new RecordInfo(getPrimaryKey(0), 0, 100));
+    recordInfoList1.add(new RecordInfo(getPrimaryKey(1), 1, 100));
+    recordInfoList1.add(new RecordInfo(getPrimaryKey(2), 2, 100));
+    recordInfoList1.add(new RecordInfo(getPrimaryKey(0), 3, 80));
+    recordInfoList1.add(new RecordInfo(getPrimaryKey(1), 4, 120));
+    recordInfoList1.add(new RecordInfo(getPrimaryKey(0), 5, 100));
+    ThreadSafeMutableRoaringBitmap validDocIds1 =
+        upsertMetadataManager.addSegment(segment1, recordInfoList1.iterator());
+    // segment1: 0 -> {0, 100}, 1 -> {4, 120}, 2 -> {2, 100}
+    checkRecordLocation(recordLocationMap, 0, segment1, 0, 100);
+    checkRecordLocation(recordLocationMap, 1, segment1, 4, 120);
+    checkRecordLocation(recordLocationMap, 2, segment1, 2, 100);
+    assertEquals(validDocIds1.getMutableRoaringBitmap().toArray(), new int[]{0, 2, 4});
+
+    // Add the second segment
+    String segment2 = SEGMENT_PREFIX + 2;
+    List<RecordInfo> recordInfoList2 = new ArrayList<>();
+    recordInfoList2.add(new RecordInfo(getPrimaryKey(0), 0, 100));
+    recordInfoList2.add(new RecordInfo(getPrimaryKey(1), 1, 100));
+    recordInfoList2.add(new RecordInfo(getPrimaryKey(2), 2, 120));
+    recordInfoList2.add(new RecordInfo(getPrimaryKey(3), 3, 80));
+    recordInfoList2.add(new RecordInfo(getPrimaryKey(0), 4, 80));
+    ThreadSafeMutableRoaringBitmap validDocIds2 =
+        upsertMetadataManager.addSegment(segment2, recordInfoList2.iterator());
+    // segment1: 0 -> {0, 100}, 1 -> {4, 120}
+    // segment2: 2 -> {2, 120}, 3 -> {3, 80}
+    checkRecordLocation(recordLocationMap, 0, segment1, 0, 100);
+    checkRecordLocation(recordLocationMap, 1, segment1, 4, 120);
+    checkRecordLocation(recordLocationMap, 2, segment2, 2, 120);
+    checkRecordLocation(recordLocationMap, 3, segment2, 3, 80);
+    assertEquals(validDocIds1.getMutableRoaringBitmap().toArray(), new int[]{0, 4});
+    assertEquals(validDocIds2.getMutableRoaringBitmap().toArray(), new int[]{2, 3});
+
+    // Replace (reload) the first segment
+    ThreadSafeMutableRoaringBitmap newValidDocIds1 =
+        upsertMetadataManager.addSegment(segment1, recordInfoList1.iterator());
+    // original segment1: 0 -> {0, 100}, 1 -> {4, 120}
```

Comment:
shall we include the removal as part of the replace? the removal of the old shall be after the addition of the new?

Reply:
Good point, added a new test for replace and remove